**Part-A**<br>

Use the Lab2 dataset provided. Load the dataset and split it into test and train. After that, train the model using Gaussian and Multinominal classifiers and post which model performs better. Use the trained model to perform some predictions on test data.

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import GaussianNB, MultinomialNB
from sklearn.metrics import accuracy_score

# Load the dataset
data = pd.read_csv('Lab2_dataset.csv')

# Split the dataset into training and testing sets
X = data['text']
y = data['label']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create a CountVectorizer
vectorizer = CountVectorizer()
X_train_vectorized = vectorizer.fit_transform(X_train)

# Train Gaussian Naive Bayes classifier
gnb = GaussianNB()
gnb.fit(X_train_vectorized.toarray(), y_train)

# Train Multinomial Naive Bayes classifier
mnb = MultinomialNB()
mnb.fit(X_train_vectorized, y_train)

X_test_vectorized = vectorizer.transform(X_test)

# Make predictions on test data
gnb_predictions = gnb.predict(X_test_vectorized.toarray())
mnb_predictions = mnb.predict(X_test_vectorized)

# Evaluate the performance
gnb_accuracy = accuracy_score(y_test, gnb_predictions)
mnb_accuracy = accuracy_score(y_test, mnb_predictions)

print("Gaussian Naive Bayes Accuracy:", gnb_accuracy)
print("Multinomial Naive Bayes Accuracy:", mnb_accuracy)


Gaussian Naive Bayes Accuracy: 0.9545893719806763
Multinomial Naive Bayes Accuracy: 0.9719806763285024


Based on the output, the Multinomial Naive Bayes model performs better than the Gaussian Naive Bayes model for the given dataset. The accuracy of the Multinomial Naive Bayes model is 0.9719, while the accuracy of the Gaussian Naive Bayes model is 0.9546.

Therefore, if accuracy is the sole criterion for determining the better-performing model, we would choose the Multinomial Naive Bayes model. However, it's important to note that accuracy alone may not always be the most reliable metric for evaluating a model's performance, especially in cases where the dataset is imbalanced or the cost of misclassification varies across different classes. It's recommended to consider other evaluation metrics and domain-specific requirements when selecting a model.

**Part-B**<br>
Use this air bnb new york city dataset and remove outliers based on price per night for a given apartment/home. You can demonstrate why using other techniques like mean/median/ percentile works. The task is to come up with a clean dataset that does not have outliers showcasing all the possibilities.


In [3]:
import pandas as pd

# Load the dataset
df = pd.read_csv("AB_NYC_2019.csv")

# Calculate the mean and standard deviation
mean_price = df['price'].mean()
std_price = df['price'].std()

# lower and upper bounds for outliers
lower_bound = mean_price - 3 * std_price
upper_bound = mean_price + 3 * std_price

# Remove outliers based on price per night
clean_df = df[(df['price'] >= lower_bound) & (df['price'] <= upper_bound)]

# you can remove outliers based on percentiles
lower_percentile = df['price'].quantile(0.01)
upper_percentile = df['price'].quantile(0.99)

clean_df = df[(df['price'] >= lower_percentile) & (df['price'] <= upper_percentile)]

# Print the first few rows of the clean dataset
print(clean_df.head())


     id                                              name  host_id   
0  2539                Clean & quiet apt home by the park     2787  \
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude   
0         John            Brooklyn    Kensington  40.64749  -73.97237  \
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review   
0     Private room    149               1                  9  20